src/gt4py/next/embedded/operators.py

```diff
@@ -0,0 +1,102 @@
+# version. See the LICENSE.txt file at the top-level directory of this
+# distribution for a copy of the license or check <https://www.gnu.org/licenses/>.
+#
+# SPDX-License-Identifier: GPL-3.0-or-later
+
+from typing import Callable, Sequence
+
+from gt4py._core import definitions as core_defs
+from gt4py.next import common, constructors, field_utils, utils
+from gt4py.next.embedded import common as embedded_common
+
+
+def scan(
+    scan_op: Callable,
+    forward: bool,
+    init: core_defs.Scalar | tuple[core_defs.Scalar | tuple, ...],
+    scan_axis: common.Dimension,
+    scan_range: common.NamedRange,
+    args: tuple,
+    kwargs: dict,
+):
+    domain_intersection = embedded_common.intersect_domains(
+        *[field_utils.get_domain(f) for f in [*args, *kwargs.values()] if _is_field_or_tuple(f)]
+    )
+    non_scan_domain = common.Domain(*[nr for nr in domain_intersection if nr[0] != scan_axis])
+
+    res_domain = common.Domain(
+        *[scan_range if nr[0] == scan_axis else nr for nr in domain_intersection]
+    )
+    if scan_axis not in res_domain.dims:
+        # even if the scan dimension is not in the input, we can scan over it
+        res_domain = common.Domain(*res_domain, (scan_range))
+
+    res = _construct_scan_array(res_domain)(init)
+
+    def combine_pos(hpos, vpos):
+        hpos_iter = iter(hpos)
+        return tuple(vpos if d == scan_axis else next(hpos_iter) for d in res_domain.dims)
+
+    def scan_loop(hpos):
+        acc = init
+        for k in scan_range[1] if forward else reversed(scan_range[1]):
+            pos = combine_pos(hpos, (scan_axis, k))
+            new_args = [_tuple_at(pos)(arg) for arg in args]
+            new_kwargs = {k: _tuple_at(pos)(v) for k, v in kwargs.items()}
+            acc = scan_op(acc, *new_args, **new_kwargs)
+            _tuple_assign(pos)(res, acc)
+
+    for hpos in embedded_common.iterate_domain(non_scan_domain):
+        scan_loop(hpos)
+    if len(non_scan_domain) == 0:
+        # if we don't have any dimension orthogonal to scan_axis, we need to do one scan_loop
+        scan_loop(())
+
+    return res
+
+
+@utils.get_common_tuple_value
+def _is_field_or_tuple(field: common.Field | tuple[common.Field, ...]) -> bool:
+    return common.is_field(field)
+
+
+def _construct_scan_array(domain: common.Domain):
+    @utils.apply_to_tuple_elements
+    def impl(init: core_defs.Scalar) -> common.Field:
+        return constructors.empty(domain, dtype=type(init))
+
+    return impl
+
+
+def _tuple_assign(
+    pos: Sequence[common.NamedIndex],
+) -> Callable[[common.MutableField | tuple[common.MutableField | tuple, ...]], None]:
+    @utils.apply_to_tuple_elements
+    def impl(target: common.MutableField, source: core_defs.Scalar):
+        target[pos] = source
+
+    return impl  # type: ignore[return-value]
+
+
+def _tuple_at(
+    pos: Sequence[common.NamedIndex],
+) -> Callable[
+    [common.Field | core_defs.Scalar | tuple[common.Field | core_defs.Scalar | tuple, ...]],
```

Comment:
```suggestion
    [common.Field | core_defs.Scalar | tuple[common.Field | core_defs.Scalar | tuple]],
```
Type annotation appears to be wrong again.

Reply:
this is correct, e.g. tuple of fields